File: typescript-sdk/src/utils/otel.ts
import { diag, DiagConsoleLogger, DiagLogLevel, Span, SpanOptions, SpanStatusCode, trace } from '@opentelemetry/api';
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { ConsoleSpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base';

let provider: NodeTracerProvider | undefined;
let disabled = false;

function ensureProvider(): void {
	if (provider || disabled) return;
	if (process.env.OTEL_DISABLED === '1') {
		disabled = true;
		return;
	}

	diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.ERROR);
	provider = new NodeTracerProvider();
	provider.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));
	provider.register();
}

export function initOtel(): void {
	ensureProvider();
}

export function getTracer() {
	ensureProvider();
	if (disabled) return undefined;
	return trace.getTracer('visa-direct-sdk');
}

type WithSpanOptions = SpanOptions & { attributes?: Record<string, unknown> };

export async function withSpan<T>(name: string, fn: (span: Span | undefined) => Promise<T> | T, options: WithSpanOptions = {}): Promise<T> {
	const tracer = getTracer();
	if (!tracer) {
		return await fn(undefined);
	}

	return await tracer.startActiveSpan(name, options, async (span) => {
		try {
			if (options.attributes) {
				for (const [key, value] of Object.entries(options.attributes)) {
					if (value !== undefined) span.setAttribute(key, value as any);
				}
			}
			const result = await fn(span);
			return result;
		} catch (err) {
			span.recordException(err as Error);
			span.setStatus({ code: SpanStatusCode.ERROR });
			throw err;
		} finally {
			span.end();
		}
	});
}

export function redact(value: unknown): string {
	if (value === undefined || value === null) return '[redacted]';
	if (typeof value === 'string') {
		if (value.includes('*')) return value;
		return '[redacted]';
	}
	return '[redacted]';
}

File: python-sdk/visa_direct_sdk/utils/otel.py
from __future__ import annotations

import os
from contextlib import contextmanager
from typing import Dict, Iterator, Optional

from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import ConsoleSpanExporter, SimpleSpanProcessor
from opentelemetry.trace import Span, Status, StatusCode

_provider_initialised = False
_disabled = False


def init_otel() -> None:
	global _provider_initialised, _disabled
	if _provider_initialised or _disabled:
		return
	if os.environ.get("OTEL_DISABLED") == "1":
		_disabled = True
		return
	provider = TracerProvider()
	provider.add_span_processor(SimpleSpanProcessor(ConsoleSpanExporter()))
	trace.set_tracer_provider(provider)
	_provider_initialised = True


def get_tracer():
	init_otel()
	return trace.get_tracer("visa-direct-sdk")


@contextmanager
def use_span(name: str, attributes: Optional[Dict[str, object]] = None) -> Iterator[Optional[Span]]:
	if _disabled or os.environ.get("OTEL_DISABLED") == "1":
		yield None
		return
	tracer = get_tracer()
	with tracer.start_as_current_span(name, attributes=attributes or {}) as span:
		try:
			yield span
		except Exception as exc:  # pragma: no cover - instrumentation
			span.record_exception(exc)
			span.set_status(Status(StatusCode.ERROR))
			raise


def redact(value: object) -> str:
	if value is None:
		return "[redacted]"
	if isinstance(value, str) and "*" in value:
		return value
	return "[redacted]"
